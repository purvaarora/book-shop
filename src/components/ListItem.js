// List item element represents a single book element in the bookshop 

import React from "react";
import PropTypes from 'prop-types'
import { FiPlusSquare, FiMinusSquare } from 'react-icons/fi'

function ListItem({
    book, // selected book to create UI element
    handleAddToCart, // handler for button to add book to cart
    handleRemoveFromCart, // handler for button to remove book from cart
    shoppingCartValue, // shoppingcart count 
    baseBookFare // book fare same for all books for this case
}) {
    const image = require('./../../public/' + book.id + '.jpeg');
    return (
        <React.Fragment>
            <img alt={book.tite} src={image.default} />
            <div className="bookStore-list-item-description">
                <div>
                    <p>{book.title}</p>
                    <p className="subText">By {book.author}</p>
                    <p>$ {parseInt(baseBookFare).toFixed(2)}</p>
                </div>
                <div className="bookStore-list-item-actionButtons">
                    <button id={`${book.id}-remove`} className="bookStore-list-item-actionButtons-remove" disabled={!shoppingCartValue} onClick={() => { handleRemoveFromCart(book.id) }}><FiMinusSquare /></button>
                    <input readOnly defaultValue={0} value={shoppingCartValue} />
                    <button id={`${book.id}-add`} className="bookStore-list-item-actionButtons-add" onClick={() => { handleAddToCart(book.id) }}><FiPlusSquare /></button>
                </div>
            </div>
        </React.Fragment>
    )
}

ListItem.propTypes = {
    book: PropTypes.shape({}).isRequired,
    handleAddToCart: PropTypes.func.isRequired,
    handleRemoveFromCart: PropTypes.func.isRequired,
    shoppingCartValue: PropTypes.number,
    baseBookFare: PropTypes.number
};

ListItem.defaultProps = {
    shoppingCartValue: 0,
    baseBookFare: 0
};

export default ListItem