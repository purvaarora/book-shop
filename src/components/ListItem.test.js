import React from 'react';
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import ListItem from './ListItem'

let initialProps = {
    book: {
        id: 'book1',
        amount: 8,
        title: 'Harry Potter and the Sorcerers Stone',
        author: 'J.K. Rowling'
    },
    handleAddToCart: jest.fn(),
    handleRemoveFromCart: jest.fn(),
    shoppingCartValue: 1
}

describe('ListItem component tests', () => {
    it('matches snapshot', () => {
        const component = renderer.create(
            <ListItem {...initialProps} />,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('call handleAddToCart on add button click', () => {
        const wrapper = shallow(<ListItem  {...initialProps}/>);
        wrapper.find('.bookStore-list-item-actionButtons-add').simulate('click');
        expect(initialProps.handleAddToCart).toHaveBeenCalledTimes(1);
    })

    it('call handleRemoveFromCart on add button click', () => {
        const wrapper = shallow(<ListItem  {...initialProps}/>);
        wrapper.find('.bookStore-list-item-actionButtons-remove').simulate('click');
        expect(initialProps.handleRemoveFromCart).toHaveBeenCalledTimes(1);
    })
})




