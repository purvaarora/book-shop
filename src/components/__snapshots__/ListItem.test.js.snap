// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component tests matches snapshot 1`] = `
Array [
  <img />,
  <div
    className="bookStore-list-item-description"
  >
    <div>
      <p>
        Harry Potter and the Sorcerers Stone
      </p>
      <p
        className="subText"
      >
        By 
        J.K. Rowling
      </p>
      <p>
        $ 
        8.00
      </p>
    </div>
    <div
      className="bookStore-list-item-actionButtons"
    >
      <button
        className="bookStore-list-item-actionButtons-remove"
        disabled={false}
        id="book1-remove"
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="3"
          />
          <line
            x1="8"
            x2="16"
            y1="12"
            y2="12"
          />
        </svg>
      </button>
      <input
        defaultValue={0}
        readOnly={true}
        value={1}
      />
      <button
        className="bookStore-list-item-actionButtons-add"
        id="book1-add"
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="3"
          />
          <line
            x1="12"
            x2="12"
            y1="8"
            y2="16"
          />
          <line
            x1="8"
            x2="16"
            y1="12"
            y2="12"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;
